docs/RFCS/20200811_non_blocking_txns.md

```diff
@@ -0,0 +1,1027 @@
+either side of its original timestamp if causally ordered before it. But of
+course, it would be impossible for the write to be causally ordered after it. So
+to ensure that all possible causal dependencies are captured and linearizability
+is enforced, a transaction will bump its timestamp to ensure that it observes
+all values in its uncertainty interval. This either incurs a transaction refresh
+or a full transaction restart.
+
+The writes performed by non-blocking transactions are no exception. When a read
+observes a write performed by a non-blocking transaction in its uncertainty
+interval, it will need to bump its timestamp so that it observes the write.
+
+The complication here is that if a reading transaction bumped its timestamp
+above a value written by a non-blocking transaction in its uncertainty interval,
+it could end up with a timestamp in the future of "present time". This could
+risk resulting in a stale read if this read immediately committed and then
+triggered a causally dependent read that hit a slower gateway where the written
+value was not considered in the second read's uncertainty interval. The following
+example demonstrates this hazard:
+
+```
+params:
+  clock_offset = 10
+  non_blocking_duration = 50
+
+clocks:
+  node A's clock = 100
+  node B's clock = 95
+
+- txn 1 begins on node A
+- txn 1 writes in future @ time 150
+- txn 1 commits
+- txn 1 begins waiting out non_blocking_duration before returning to client
+
+- time advances by 40
+clocks:
+  node A's clock = 140
+  node B's clock = 135
+
+- txn 2 begins on node A
+- txn 2 reads at 140 and observes write in uncertainty interval
+- txn 2 returns write to client
+- txn 2 commits and acks client
+
+- txn 3 is triggered by commit of txn 2 (causal dependency)
+- txn 3 begins on node B
+- txn 3 reads at 135 and observed no write in uncertainty interval
+- txn 3 returns nothing to client
+- txn 3 commits and acks client
+
+- time advances by 10
+clocks:
+  node A's clock = 150
+  node B's clock = 145
+
+- txn1 acks client
+
+HAZARD! txn3 just saw stale read. It was triggered by txn2, which had observed
+        the value written by txn1, but it did not observe the value.
+```
+
+Careful readers may notice that the first read-only transaction (txn2) would
+have had to bump its timestamp to 150 upon observing a value in its uncertainty
+interval. So if it had respected the Commit Wait rule then it would have been
+delayed before committing, after which point the stale read would not be
+possible because node B's clock would have advanced far enough that the value
+would have been in the second read-only transaction's (txn3) uncertainty
+interval.
+
+But relying on Commit Wait here violates a desirable property of our transaction
+model – that committing a read-only transaction is a no-op. To recover this
+property, we'll want to wait out the uncertainty before the refresh/retry.
+Again, the easiest way to do this is to `hlc.Update(txn.ReadTimestamp)` before
```

Comment:
Maybe this relates to @tbg's concern above, but doesn't this mean that `txn2` would have a higher latency than a normal transaction that performs a read refresh?
-- Edit: I think you covered this in "Non-blocking and Standard Transaction Interactions", but might be worth adding a pointer here.